pkg/registry/core/shoot/storage/admin_kubeconfig.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	shootStateObj, err := r.shootStateStorage.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	shootState := &corev1alpha1.ShootState{}
+
+	err = kubernetes.GardenScheme.Convert(shootStateObj, shootState, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	shootObj, err := r.shootStorage.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	shoot, ok := shootObj.(*core.Shoot)
+	if !ok {
+		return nil, errors.NewInternalError(fmt.Errorf("cannot convert to *core.Shoot object - got type %T", shootObj))
+	}
+
+	if len(shoot.Status.AdvertisedAddresses) == 0 {
+		return nil, errors.NewBadRequest("no kube-apiserver advertised addresses in Shoot .status.advertisedAddresses")
+	}
+
+	var caData []byte
+
+	ca, err := infodata.GetInfoData(shootState.Spec.Gardener, corev1beta1constants.SecretNameCACluster)
+	if err != nil {
+		return nil, errors.NewInternalError(err)
+	}
+
+	if ca == nil {
+		return nil, errors.NewBadRequest("certificate authority not yet provisioned")
+	}
+
+	caInfoData, ok := ca.(*secrets.CertificateInfoData)
+	if !ok {
+		return nil, errors.NewInternalError(fmt.Errorf("could not convert InfoData entry ca to CertificateInfoData"))
+	}
+
+	caCert, err := secrets.LoadCertificate("", caInfoData.PrivateKey, caInfoData.Certificate)
+	if err != nil {
+		return nil, errors.NewInternalError(err)
+	}
+
+	now := r.now()
+	cc := &certConfig{
+		subject: pkix.Name{
+			CommonName:   user.GetName(),
+			Organization: []string{"system:masters"},
+		},
+		now:        now,
+		expiration: now.Add(time.Duration(out.Spec.ExpirationSeconds) * time.Second),
+		caCert:     caCert.Certificate,
+		caKey:      caCert.PrivateKey,
+	}
+
+	clientCert, clientKey, err := cc.newKeyPair()
+	if err != nil {
+		return nil, err
+	}
+
+	caData = caInfoData.Certificate
+	authName := fmt.Sprintf("%s--%s", shoot.Namespace, shoot.Name)
+
+	config := &configv1.Config{
+		Clusters: []configv1.NamedCluster{},
+		Contexts: []configv1.NamedContext{},
+		AuthInfos: []configv1.NamedAuthInfo{{
+			Name: authName,
+			AuthInfo: configv1.AuthInfo{
+				ClientCertificateData: clientCert,
+				ClientKeyData:         clientKey,
+			},
+		}},
+	}
+
+	for i, address := range shoot.Status.AdvertisedAddresses {
+
+		configName := fmt.Sprintf("%s-%s", authName, address.Name)
+		config.Clusters = append(config.Clusters, configv1.NamedCluster{
+			Name: configName,
+			Cluster: configv1.Cluster{
+				CertificateAuthorityData: caData,
+				Server:                   address.URL,
+			},
+		})
+		config.Contexts = append(config.Contexts, configv1.NamedContext{
+			Name: configName,
+			Context: configv1.Context{
+				Cluster:  configName,
+				AuthInfo: authName,
+			},
+		})
+
+		if i == 0 {
+			config.CurrentContext = configName
+		}
+	}
+
+	configData, err := runtime.Encode(configlatest.Codec, config)
+	if err != nil {
+		return nil, err
+	}
+
+	out.Status.KubeConfig = []byte(configData)
```

Comment:
Why don't you use https://github.com/gardener/gardener/blob/master/pkg/utils/secrets/control_plane.go for this? It already should have everything you need?

> The kubeconfig uses NIST P-256 elliptic curve for client certificate for authentication.

Why/what for?

Reply:
1. I want to switch all certificates to ECC (elliptic curve cryptography) to the lower key size. 
2. At the moment it doesn't offer expiration time for the certificates.
3. It doesn't offer ECC
4. The kubeconfigs only have one context and cluster, in our case we want more

In general yes, we should do this, but my suggestion is to move this refactoring to an separate PR.

Reply:
Wouldn't it be better to enhance the secret util package already now so that it supports your missing features (expiration time and multiple contexts for the kubeconfig)? You can also even start with RSA or make it configurable if required.

From my understanding, it's not a refactoring to enhance the utility library with new features if you require them for preventing code duplication. It'd indeed be a refactoring later if we merge as is.
Unless there is a good argument for proceeding as proposed, I'd vote for rather doing it now as you are already on it (we all know that once code is merged then the burden is much higher to touch it again).

Reply:
Quite the opposite - refactoring and adding new functionality should be separate activities (PRs). If there both are added in a same PR and there is a bug in the refactoring or in the new functionality then revert would also remove the working parts.

Also the PR would grow too large to manage / review.

Reply:
After this PR is in I would start working on the refactoring. I'll even do it in stages (PRs):

- first the EC certificates
- the kubeconfig changes

Small changes are quite easiliy to review.

Reply:
OK, then what about doing it vice-versa, i.e., first enhance/"refactor" the secret utils library and then rebase this PR on top of it? This way, it will be even easier  to review and less risk because all the changes you introduce here would not exist in the first place.